packages/compose/src/hooks/use-merge-refs/index.js

```diff
@@ -0,0 +1,23 @@
+/**
+ * External dependencies
+ */
+import mergeRefs from 'react-merge-refs';
+
+/**
+ * WordPress dependencies
+ */
+import { useCallback } from '@wordpress/element';
+
+/** @typedef {import('@wordpress/element').RefObject} RefObject */
+
+/**
+ * Merges the given refs into a new ref, which will persist for the full
+ * lifetime of the component, as a ref should be.
+ *
+ * @param  {...RefObject|Function} refs Ref objects or callbacks to merge.
+ *
+ * @return {Function} A new ref callback.
+ */
+export default function useMergeRefs( ...refs ) {
+	return useCallback( mergeRefs( refs ), refs );
```

Comment:
Actually, thinking more, the addition of "refs" as a dependency is not a good thing.  If one of the refs change, all the refs will be recalled again even the ones that didn't change. Not sure exactly how to solve this. We could solve this by "convention".

Meaning we make sure we only use `useMergeRefs` for refs that don't change over time (so even the useCallback we use right now for these refs is not necessary. It  doesn't sound great though, so I'm wondering if there's a smarter solution (only call the ref  function again if it changes). I also wonder whether such thing exists in npm already.

Reply:
I agree, I'm not sure if it a good thing, because on the one hand there's the assumption that a ref callback shouldn't be called if there's no **new** node, and on the other hand, there's the dependency issue, which you've resolved before by setting the dependencies as refs.

I think we should always return the same callback throughout the lifecycle as refs are supposed to behave.

Reply:
Another way is to keep using `mergeRefs` like today and just "memoize" the ref callbacks. I think it might be the best way. We could make it a hook, useMemoizeCallback or something like that :P

Reply:
You mean replace `useCallback` with `useMemo`?

Reply:
no, I mean `const ref = useCallback( memize( (node) => { // do something } ), [ deps ] )`

Reply:
I'm thinking more and more that, this is the best approach. we could name this `useMemoizedRef( callback, debs )`

Reply:
Hm, I'm not following. `useCallback` will already memoize it?

Reply:
yes but not enough :P 

the memize unsure the callback is not called again unless `node` changes.

Reply:
we could also apply a maxSize of `1` to the memize call as we only care about the previous call.

Reply:
But why not prevent the call entirely?

Reply:
One of the issues is that you wouldn't expect the ref callback to be called if nothing changed.

Reply:
Not sure I understand, memize prevents the call if "node" doesn't change which is what we want?

Reply:
Yes, that's what we want, but now we are adjusting to a ref callback not being "purely" called when the node changes (or is added or removed). To me the expectation I have with a ref callback is that it's never called multiple times with the same node.

Reply:
It seems to me "purely" called doesn't mean much. useMemo or useCallback do the same thing as memize behind the scenes. I don't really see the difference. The difference is in one case, the memorization is inside React itself while in the other case it's outside React.

Reply:
You're right

Reply:
I still think there's benefit in wrapping the function in `useCallback`, and add the refs as dependencies, which makes the makes all the inner `useCallback`s work correctly. What's the point otherwise of wrapping ref callbacks in `useCallback`?

Reply:
the point is that the a  new function is generated for the ref when one of the deps change which calls the callback again with the node, allowing you to react properly to these deps changes.